### PR TITLE
Add TagInput component

### PR DESCRIPTION
**Pull request by renaudtertrais**
tag-input

### CODE REVIEW

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+import { bemClass } from '../../helpers/bem';
+
+import useEventListener from '../../hooks/useEventListener';
+import useFocusInOut from '../../hooks/useFocusInOut';
+import Tag from '../../Tag';
+
+import './TagInput.scss';
+
+const sliceArr = (xs, i, j) =>
```
Comment (davidjomain) [inline src/alto-ui/Form/TagInput/TagInput.js]: Maybe you can extract this as an helper ?

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+
+    const isTagEqual = a => b => getTagValue(a) === getTagValue(b);
+    const isTagDiff = a => b => !isTagEqual(a)(b);
+    const isTagActive = tag => selection.some(isTagEqual(tag));
+
+    const availableTags = props.availableTags(value);
+    const remainingTags = (availableTags || []).filter(t => !tags.some(isTagEqual(t)));
+    const match =
+      props.match ||
+      (v =>
```
Comment (NicolasMartos) [inline src/alto-ui/Form/TagInput/TagInput.js]: `v` for ?

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        availableTags
+          ? remainingTags.some(t => getTagValue(t) === v)
+          : !tags.some(t => getTagValue(t) === v));
+
+    const addToSelection = (...xs) =>
```
Comment (NicolasMartos) [inline src/alto-ui/Form/TagInput/TagInput.js]: Why not 
```suggestion
    const addToSelection = x) =>
```

File: src/alto-ui/Form/TagInput/TagInput.js (hunk 2 above)
Comment (davidjomain) [reply src/alto-ui/Form/TagInput/TagInput.js]: value ?

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+    const instance = useRef({
+      keyupOnDocumentListener(e) {
+        if (e.key === 'Backspace' || e.key === 'Delete') {
+          instance.removeTagsInSelection();
+          setSelection([]);
+          setPosition(null);
+        }
+        if (e.key === 'Escape') {
+          e.nativeEvent.stopImmediatePropagation();
```
Comment (NicolasMartos) [inline src/alto-ui/Form/TagInput/TagInput.js]: Need to use nativeEvent ?

File: src/alto-ui/Form/TagInput/TagInput.js (hunk 2 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.js]: vendetta ?

Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.js]: value

File: src/alto-ui/Form/TagInput/TagInput.js (hunk 4 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.js]: https://gist.github.com/ggregoire/ce7bc946212920c0a6bad8125567d001

File: src/alto-ui/Form/TagInput/TagInput.scss [hunk 5]
```diff
@@ -0,0 +1,49 @@
+@import '../../scss/inc';
+
+.TagInput__tags-list {
+  display: flex;
```
Comment (djdeimon) [inline src/alto-ui/Form/TagInput/TagInput.scss]: Component height "jumps" when user enter the tag

input_tag

File: src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  const isActive = useFocusInOut(null, mainRef);
+
+  const state = {
+    value: useResettableState(''),
+    selection: useResettableState([]),
+    position: usePosition(tags),
+    isActive,
+  };
+
+  const [value, , resetValue] = state.value;
```
Comment (djdeimon) [inline src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js]: Why not return object from `useResettableState` and make reference with by simply `.` 
You can avoid that kind of strange destructurization assigment when you must put empty `,` to skip value.  ?

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+      <div ref={refs.main} className={bemClass('TagInput', {}, className)}>
+        <div className="TagInput__tags-list TagInput__tags-list--value" ref={refs.value}>
+          <button
+            disabled={disabled}
+            className="TagInput__button"
+            onClick={() => setPosition(null)}
+          />
+          {tags.map((tag, index) => (
+            <Fragment key={getTagValue(tag)}>
+              {position === index ? input : <div style={{ width: 1 }} />}
```
Comment (djdeimon) [inline src/alto-ui/Form/TagInput/TagInput.js]: `<div style={{ width: 1 }} />` maybe some class instead set here style ? `TagInput__carret` ?

File: src/alto-ui/Form/TagInput/TagInput.js [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+        </div>
+        {!disabled && state.isActive && !!remainingTags.length && (
+          <div className="TagInput__tags-list">
+            {remainingTags.map(tag => (
+              <div key={getTagValue(tag)} className="TagInput__tag">
+                {renderTag(tag, {
+                  disabled: !getTagValue(tag).includes(value),
+                  onClick: () => {
+                    handleNewTag(tag);
+                    setTimeout(() => setPosition(null), 0);
```
Comment (djdeimon) [inline src/alto-ui/Form/TagInput/TagInput.js]: `setTimeout(() => setPosition(null), 0);` neccessary ?

File: src/alto-ui/Form/TagInput/TagInput.scss (hunk 5 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.scss]: yes I know. Not a big issue right, will fix it if needed later.

File: src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js (hunk 6 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js]: very good point, I am not a big fan also. I tried to keep the React style. But using an object instead was actually my first dev. Check this commit and tell me if you prefer. https://github.com/beqom/alto/pull/564/commits/aa28db42331efbdb0310bf440a17414999a5b2fa

File: src/alto-ui/Form/TagInput/TagInput.js (hunk 7 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.js]: yep you' re right

File: src/alto-ui/Form/TagInput/TagInput.js (hunk 8 above)
Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/TagInput.js]: I think the idea was to pospone the setState after the render (at the end of the stack). I will check if it is mandatory.

File: src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js (hunk 6 above)
Comment (djdeimon) [reply src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js]: Yep, I know that the main point to use array is that you can name variables as you want but maybe instead writing 
```
  const [position, , resetPosition] = state.position;
```
we can use aliases
```
const {get: position, reset: resetSelecion} = state.position;
```
And this is more obvious than some indexes skipping

Comment (renaudtertrais) [reply src/alto-ui/Form/TagInput/hooks/useTagInputStateAndRefs.js]: we shared the same analysis with @NicolasMartos . I will do the second one ;)

**Pull request merged**